### PR TITLE
[KAFKA-4468] Correctly calculate the window end timestamp after read from state stores

### PR DESCRIPTION
**Pull request by ConcurrencyPractitioner**
I have decided to use the following approach to fixing this bug:

1) Since the Window Size in WindowedDeserializer was originally unknown, I have initialized
a field _windowSize_ and created a constructor to allow it to be instantiated

2) The default size for __windowSize__ is _Long.MAX_VALUE_. If that is the case, then the 
deserialize method will return an Unlimited Window, or else will return Timed one.

3) Temperature Demo was modified to demonstrate how to use this new constructor, given
that the window size is known.

### CODE REVIEW

**Comment by bbejeck** [comment]
@ConcurrencyPractitioner thanks for the PR.  Can you address the checkstyle errors in the build? Will review after that.  Thanks.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java [hunk 1]
```diff
@@ -64,13 +72,15 @@ public void configure(Map<String, ?> configs, boolean isKey) {
     public Windowed<T> deserialize(String topic, byte[] data) {
 
         byte[] bytes = new byte[data.length - TIMESTAMP_SIZE];
-
+        
+        System.out.println("deserialize method cDs.alled");
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: I'm guessing this is left over from debugging, can you take this out?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java [hunk 2]
```diff
@@ -35,14 +36,21 @@
     private static final int TIMESTAMP_SIZE = 8;
 
     private Deserializer<T> inner;
-
+    private Long windowSize;
     // Default constructor needed by Kafka
     public WindowedDeserializer() {}
 
     public WindowedDeserializer(Deserializer<T> inner) {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: nit: make parameters final

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java [hunk 3]
```diff
@@ -35,14 +36,21 @@
...
         this.inner = inner;
+        this.windowSize = Long.MAX_VALUE;
     }
 
+    public WindowedDeserializer(Deserializer<T> inner, long windowSize){
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: nit: same as above for parameters

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java [hunk 4]
```diff
@@ -64,13 +72,15 @@ public void configure(Map<String, ?> configs, boolean isKey) {
...
+        System.out.println("deserialize method cDs.alled");
         System.arraycopy(data, 0, bytes, 0, bytes.length);
-
+        
         long start = ByteBuffer.wrap(data).getLong(data.length - TIMESTAMP_SIZE);
-
-        // always read as unlimited window
-        return new Windowed<T>(inner.deserialize(topic, bytes), new UnlimitedWindow(start));
+        
+        return windowSize != Long.MAX_VALUE ? new Windowed<T>(inner.deserialize(topic, bytes),
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: IMHO a little hard to follow and some logic is duplicated,  what about something like
```
Window timeWindow = windowSize != Long.MAX_VALUE ? WindowStoreUtils.timeWindowForSize(start, windowSize) : new UnlimitedWindow(start);

return new Windowed<T>(inner.deserialize(topic, bytes), timeWindow);
```

**Comment by bbejeck** [comment]
@ConcurrencyPractitioner thanks again, I've left some initial comments.  
Also can we get a unit test for `WindowedDeserializer` and include the (presently unused) constructor on line 43?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java (hunk 1 above)
Comment (ConcurrencyPractitioner) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: That was the left overs from my attempts, and I probably was sloppy when removing vestiges of my work.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java (hunk 4 above)
Comment (ConcurrencyPractitioner) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: It should be noted that the Window type will also have to be imported, since it was not used
anywhere else in the code. I would use TimeWindow type instead for declaration.

**Comment by ConcurrencyPractitioner** [comment]
@bbejeck  
I have resolved your requests with the code I have just uploaded.
A unit test was also added.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 5]
```diff
@@ -144,4 +144,31 @@ public void testWindowedDeserializerNoArgConstructors() {
         windowedDeserializer.close();
         windowedDeserializer1.close();
     }
+    
+    @Test
+    public void testWindowedDeserializerWindowSize() {
+        Map<String, String> props = new HashMap<>();
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: nit: final

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 6]
```diff
@@ -144,4 +144,31 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        // test key[value].deserializer.inner.class takes precedence over serializer.inner.class
+        WindowedDeserializer<StringSerializer> windowedDeserializer = new WindowedDeserializer<>();
+        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "host:1");
+        props.put(StreamsConfig.APPLICATION_ID_CONFIG, "appId");
+        props.put("key.deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        windowedDeserializer.configure(props, true);
+        Deserializer<?> inner = windowedDeserializer.innerDeserializer();
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: ditto

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 7]
```diff
@@ -144,4 +144,31 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        long randomLong = 5000000;
+        byte[] byteValues = stringSerializer.serialize(topicName, "dummy string");
+        WindowedDeserializer<StringSerializer> windowedDeserializer1 = 
+            new WindowedDeserializer<>(windowedDeserializer.innerDeserializer(), randomLong);
+        windowedDeserializer1.configure(props, false);
+        Windowed<?> windowed = windowedDeserializer1.deserialize(topicName, byteValues);
+        long actualSize = windowed.window().end() - windowed.window().start();
+        assertEquals(randomLong, actualSize);
+        windowedDeserializer.close();
+        windowedDeserializer1.close();
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: ditto for the rest of the test

**Comment by bbejeck** [comment]
@ConcurrencyPractitioner thanks for the updates, just a few additional minor comments

**Comment by ConcurrencyPractitioner** [comment]
The Java 8 build check timed out. Could some administrator retrigger the test?

Thanks.

**Comment by bbejeck** [comment]
retest this please

**Comment by guozhangwang** [comment]
retest this please

**Comment by ConcurrencyPractitioner** [comment]
@bbejeck Test has been recompiled. All checks have passed.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 8]
```diff
@@ -144,4 +144,31 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        assertNotNull("Inner deserializer should be not null", inner);
+        assertTrue("Inner deserializer type should be StringDeserializer", inner instanceof StringDeserializer);
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+        //test for deserializer expected window end time
+        final long randomLong = 5000000;
+        byte[] byteValues = stringSerializer.serialize(topicName, "dummy string"); //dummy string, serves no real purpose
+        final WindowedDeserializer<StringSerializer> windowedDeserializer1 = 
+            new WindowedDeserializer<>(windowedDeserializer.innerDeserializer(), randomLong);
+        windowedDeserializer1.configure(props, false);
+        Windowed<?> windowed = windowedDeserializer1.deserialize(topicName, byteValues);
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: super nit:  `byteValues` on 164 and `windowed` on 168 should be final as well

**Comment by bbejeck** [comment]
@ConcurrencyPractitioner just one last minor comment, otherwise LGTM

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java [hunk 9]
```diff
@@ -33,16 +34,26 @@
...
-
+    private final Long windowSize;
+    
     private Deserializer<T> inner;
-
+    
     // Default constructor needed by Kafka
-    public WindowedDeserializer() {}
+    public WindowedDeserializer() {
+        this.windowSize = Long.MAX_VALUE;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: `this(null)` or `this(null, Long.MAX_VALUE)`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java [hunk 10]
```diff
@@ -33,16 +34,26 @@
 public class WindowedDeserializer<T> implements Deserializer<Windowed<T>> {
 
     private static final int TIMESTAMP_SIZE = 8;
-
+    private final Long windowSize;
+    
     private Deserializer<T> inner;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java:40]: add `final` (to put an additional improvement)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java [hunk 11]
```diff
@@ -33,16 +34,26 @@
...
+    
     // Default constructor needed by Kafka
-    public WindowedDeserializer() {}
+    public WindowedDeserializer() {
+        this.windowSize = Long.MAX_VALUE;
+    }
 
     public WindowedDeserializer(Deserializer<T> inner) {
         this.inner = inner;
+        this.windowSize = Long.MAX_VALUE;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: replace both lines with `this(inner, Long.MAX_VALUE)`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java [hunk 12]
```diff
@@ -33,16 +34,26 @@
...
     }
 
+    public WindowedDeserializer(Deserializer<T> inner, final long windowSize) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: - add `final` to `inner`
- code formatting: use single line per parameter

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 13]
```diff
@@ -144,4 +144,31 @@ public void testWindowedDeserializerNoArgConstructors() {
         windowedDeserializer.close();
         windowedDeserializer1.close();
     }
+    
+    @Test
+    public void testWindowedDeserializerWindowSize() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: This method tests multiple things at once and thus should be split into multiple methods. Also, use self describing names for the methods. For example for the first test: `shouldWrapUserSerdeFromConfig()` or similar.

**Review by mjsax — CHANGES_REQUESTED** [review]
Thanks for the PR @ConcurrencyPractitioner !

I think we should have an integration test for this, too.

**Comment by guozhangwang** [comment]
@ConcurrencyPractitioner Thanks for the PR.

Just to clarify on the scope of the original JIRA, which is to fix the issue that the windowed key that is read from the state store does not calculate the end timestamp. However the newly added API

```
public WindowedDeserializer(Deserializer<T> inner, final long windowSize)
```

was not yet used in the callers yet so the JIRA itself is still not resolved. Do you have plans to do that in a follow-up PR?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java (hunk 10 above)
Comment (ConcurrencyPractitioner) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java:40]: The final keyword cannot be attached to inner because inner would also be reassigned in the reconfig method.

**Comment by ConcurrencyPractitioner** [comment]
This pull request should lay the foundations for resolving this issue. I would think about doing a follow on where the callers will use the constructor that I made.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java (hunk 10 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java:40]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java [hunk 14]
```diff
@@ -33,16 +34,26 @@
...
+        this(null, Long.MAX_VALUE);
+    }
 
     public WindowedDeserializer(Deserializer<T> inner) {
-        this.inner = inner;
+        this(inner, Long.MAX_VALUE);
     }
 
+    public WindowedDeserializer(final Deserializer<T> inner,
+            final long windowSize) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: Nit: fix indention. `final` keywords should align

**Review by mjsax — COMMENTED** [review]
Please split the test into multiple ones. Thx. Otherwise LGTM.

**Comment by ConcurrencyPractitioner** [comment]
@mjsax Could you explain how I would split it into multiple tests? Do I just trim it?

**Comment by mjsax** [comment]
You make multiple methods, and each test method tests one property. See `TopologyTest` for an example. A single test methods, ends with an assertion... You could brake it into 3 methods for you case: `shouldWrapUserSerdeFromConfig()`, `shouldSerializeWindowWithSize()` and `shouldDeserializeWindowWithSize()`. Maybe like into `SessionKeySerdeTest`, too.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 15]
```diff
@@ -143,7 +144,48 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        final WindowedDeserializer<StringSerializer> windowedDeserializer = new WindowedDeserializer<>();
+        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "host:1");
+        props.put(StreamsConfig.APPLICATION_ID_CONFIG, "appId");
+        props.put("key.deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        windowedDeserializer.configure(props, true);
+        final Deserializer<?> inner = windowedDeserializer.innerDeserializer();
+        assertNotNull("Inner deserializer should be not null", inner);
+        assertTrue("Inner deserializer type should be StringDeserializer", inner instanceof StringDeserializer);
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.ByteArraySerializer");
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: This line can be remove?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 16]
```diff
@@ -143,7 +144,48 @@ public void testWindowedDeserializerNoArgConstructors() {
         windowedDeserializer1.close();
     }
     @Test
-    public void testWindowedDeserializerWindowSize() {
+    public void testWindowSerializeConfig() {
+        final Map<String, String> props = new HashMap<>();
+        // test key[value].serializer.inner.class takes precedence over serializer.inner.class
+        final WindowedDeserializer<StringSerializer> windowedDeserializer = new WindowedDeserializer<>();
+        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "host:1");
+        props.put(StreamsConfig.APPLICATION_ID_CONFIG, "appId");
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: I don't think we need to set bootstrap server and application.id for this test?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 17]
```diff
@@ -143,7 +144,48 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        props.put("key.deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: There is not reason to specify this twice. If you want to test that it work for both parameter names, it should be two different test, each testing one parameter name

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 18]
```diff
@@ -143,7 +144,48 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        windowedDeserializer.configure(props, true);
+        final Deserializer<?> inner = windowedDeserializer.innerDeserializer();
+        assertNotNull("Inner deserializer should be not null", inner);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: `null` check is redundant as `null instanceof StringDeserializer` will return false anyway.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 19]
```diff
@@ -143,7 +144,48 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+    public void testWindowSerializeWithWindowSize() {
+        final Map<String, String> props = new HashMap<>();
+        // test key[value].serializer.inner.class takes precedence over serializer.inner.class
+        final WindowedDeserializer<StringSerializer> windowedDeserializer = new WindowedDeserializer<>();
+        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "host:1");
+        props.put(StreamsConfig.APPLICATION_ID_CONFIG, "appId");
+        props.put("key.deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        windowedDeserializer.configure(props, true);
+        final Deserializer<?> inner = windowedDeserializer.innerDeserializer();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: you don't use `windowedDeserializer ` or `inner` below -- can be removed IMHO

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 20]
```diff
@@ -155,7 +197,21 @@ public void testWindowedDeserializerWindowSize() {
         final Deserializer<?> inner = windowedDeserializer.innerDeserializer();
         assertNotNull("Inner deserializer should be not null", inner);
         assertTrue("Inner deserializer type should be StringDeserializer", inner instanceof StringDeserializer);
+    }
+    @Test
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: Nit: add newline

**Review by mjsax — COMMENTED** [review]
Thanks for the update.

Some more comments about the test. I guess there is more cleanup you can do there.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 21]
```diff
@@ -138,10 +138,88 @@ public void testWindowedDeserializerNoArgConstructors() {
...
         assertNotNull("Inner deserializer should be not null", inner1);
         assertTrue("Inner deserializer type should be ByteArrayDeserializer", inner1 instanceof ByteArrayDeserializer);
         windowedDeserializer.close();
         windowedDeserializer1.close();
     }
+    
+    @Test
+    public void testWindowSerializeConfig() {
+        final Map<String, String> props = new HashMap<>();
+        // test key[value].serializer.inner.class takes precedence over serializer.inner.class
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: If this should be tested, it should be part of the test name -> `shouldUseNewConfigNameForKeySeralizer` -> if the test has this name, you can remove this comment as the test name explains the test.

also, you should specify an invalid class name for `deserializer.inner.class` -- if both are the same, the test cannot check if the right one is picked.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 22]
```diff
@@ -138,10 +138,88 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        final WindowedDeserializer<StringSerializer> windowedDeserializer = new WindowedDeserializer<>();
+        props.put("key.deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        windowedDeserializer.configure(props, true);
+        final Deserializer<?> inner = windowedDeserializer.innerDeserializer();
+        assertTrue("Inner deserializer type should be StringDeserializer", inner instanceof StringDeserializer);
+    }
+    
+    @Test
+    public void testWindowSerializeWithWindowSize() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: test name could be better -- what property is tested?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 23]
```diff
@@ -138,10 +138,88 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        windowedDeserializer1.configure(props, false);
+        final Windowed<?> windowed = windowedDeserializer1.deserialize(topicName, byteValues);
+        final long actualSize = windowed.window().end() - windowed.window().start(); //find actual window time
+        assertEquals(randomLong, actualSize); //testing if window size matches up with expected one
+        windowedDeserializer.close();
+        windowedDeserializer1.close();
+    }
+    
+    @Test
+    public void testWindowDeserializeConfig() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: What does this actually test differently as above?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 24]
```diff
@@ -138,10 +138,88 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        props.put("key.deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        windowedDeserializer.configure(props, true);
+        final Deserializer<?> inner = windowedDeserializer.innerDeserializer();
+        assertNotNull("Inner deserializer should be not null", inner);
+        assertTrue("Inner deserializer type should be StringDeserializer", inner instanceof StringDeserializer);
+    }
+    
+    @Test
+    public void testWindowDeserializeWithWindowSize() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: as above

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 25]
```diff
@@ -138,10 +138,88 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        windowedDeserializer.configure(props, true);
+        final Deserializer<?> inner = windowedDeserializer.innerDeserializer();
+        assertTrue("Inner deserializer type should be StringDeserializer", inner instanceof StringDeserializer);
+    }
+    
+    @Test
+    public void testWindowSerializeWithWindowSize() {
+        final Map<String, String> props = new HashMap<>();
+        // test key[value].serializer.inner.class takes precedence over serializer.inner.class
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: this is tested in first test method already, and can be removed. (not just the comment, but also the corresponding code.)

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 26]
```diff
@@ -138,10 +138,88 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        final WindowedDeserializer<StringSerializer> windowedDeserializer = new WindowedDeserializer<>();
+        props.put("key.deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        windowedDeserializer.configure(props, true);
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.ByteArraySerializer");
+        props.remove("key.serializer.inner.class");
+        props.remove("value.serializer.inner.class");
+        //test for serializer expected window end time
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: seems actual test start here? if yes, remove everything from above. It seems test name should be `shouldReturnWindowWithCorrectSizeWhenDeserializing` ?

**Review by mjsax — COMMENTED** [review]
Some more comments.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java (hunk 26 above)
Comment (ConcurrencyPractitioner) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: The deserializer.inner.class in props map could not be removed, or otherwise that will result in test failure. 
Description of the failure:
```
org.apache.kafka.streams.kstream.internals.WindowedStreamPartitionerTest > testWindowSerializeWithWindowSize FAILED
    java.lang.ClassCastException: class org.apache.kafka.common.serialization.ByteArraySerializer
        at java.lang.Class.asSubclass(Class.java:3404)
        at org.apache.kafka.common.utils.Utils.newInstance(Utils.java:308)
        at org.apache.kafka.streams.kstream.internals.WindowedDeserializer.configure(WindowedDeserializer.java:66)
        at org.apache.kafka.streams.kstream.internals.WindowedStreamPartitionerTest.testWindowSerializeWithWindowSize(WindowedStreamPartitionerTest.java:166) ```

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: Makes sense. There is still some cleanup to do.... `windowedDeserializer` is not use anymore in the test but only `windowedDeserializer1` -- also, setting up `props` has `put` and later `remove` that can both be removed...

When cleaning up: consider what is *required* to set up the test, and what is not. Remove everything that is not required.

Comment (ConcurrencyPractitioner) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: The arrangement adopted from the first test method could not be changed. Removing the properties will result in the following exception: 

```
org.apache.kafka.streams.kstream.internals.WindowedStreamPartitionerTest > testWindowSerializeExpectedWindowSize FAILED
    java.lang.NullPointerException
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at org.apache.kafka.common.utils.Utils.newInstance(Utils.java:308)
        at org.apache.kafka.streams.kstream.internals.WindowedDeserializer.configure(WindowedDeserializer.java:67)
        at org.apache.kafka.streams.kstream.internals.WindowedStreamPartitionerTest.testWindowSerializeExpectedWindowSize(WindowedStreamPartitionerTest.java:154)
```

Comment (ConcurrencyPractitioner) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: @mjsax Also, the windowedDeserializer's inner class is needed for instantiation in windowedDeserializer1. It could not be removed.

**Comment by ConcurrencyPractitioner** [comment]
@mjsax I have taken care of your suggestions.

**Comment by guozhangwang** [comment]
Looks promising to me overall. @ConcurrencyPractitioner regarding the unit tests comments that @mjsax made, I think the main idea is that 1) we want to make each unit test case to be modular (that's why he suggested to break the original big test into smaller ones) and 2) each unit test function name would better be self-describing (e.g. shouldDoThisUnderThatCondition), and 3) we want to reduce unnecessary and duplicated code as much as possible if that can be shared among multiple test cases. You can take a look at the coding guide-lines here: https://github.com/apache/kafka-site/blob/asf-site/coding-guide.html#L106

For example, in the `testWindowSerializeExpectedWindowSize` function the object `windowedDeserializer` is not used, only `windowedDeserializer1` is used. So this object can be removed. Generally we want to keep each test case as succinct and simple as possible.

**Comment by ConcurrencyPractitioner** [comment]
Retest this please.

**Comment by ConcurrencyPractitioner** [comment]
@mjsax The tests have been simplified, their is only one windowedDeserializer instantiated in each test (the ones I wrote).

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 27]
```diff
@@ -138,10 +137,42 @@ public void testWindowedDeserializerNoArgConstructors() {
...
         windowedDeserializer1.close();
     }
+    
+    @Test
+    public void testWindowSerializeExpectedWindowSize() {
+        final long randomLong = 5000000;
+        final Map<String, String> props = new HashMap<>();
+        final WindowedDeserializer<StringSerializer> windowedDeserializer = new WindowedDeserializer<>(randomLong);
+        props.put("key.deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
+        props.put("deserializer.inner.class", "org.apache.kafka.common.serialization.StringDeserializer");
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: Isn't this property ignored if you set `key.deserializer.inner.class` above already? Guess this line can be removed?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java [hunk 28]
```diff
@@ -138,10 +137,42 @@ public void testWindowedDeserializerNoArgConstructors() {
...
+        //test for serializer expected window end time
+        final byte[] byteValues = stringSerializer.serialize(topicName, "dummy string"); //dummy string, serves no real purpose
+        final Windowed<?> windowed = windowedDeserializer.deserialize(topicName, byteValues);
+        final long actualSize = windowed.window().end() - windowed.window().start(); //find actual window time
+        assertEquals(randomLong, actualSize); //testing if window size matches up with expected one
+        windowedDeserializer.close();
+    }
+    
+    @Test
+    public void testWindowDeserializeExpectedWindowSize() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java]: It seems that this test does exactly the same thing as the one above? Or did I miss anything?

**Review by mjsax — COMMENTED** [review]
Thanks for the update.

**Comment by ConcurrencyPractitioner** [comment]
Retest this please.

**Comment by ConcurrencyPractitioner** [comment]
@mjsax Tests have been reduced to one, redundant one removed.

**Comment by guozhangwang** [comment]
retest this please.

**Review by mjsax — APPROVED** [review]
Thanks for the PR. Call for final review and merging @guozhangwang 

(Reminder: this does not resolve KAFKA-4468 yet, but is only one step close to a fix. Jira should not be closed.

**Comment by mjsax** [comment]
retest this please.

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk, thanks @ConcurrencyPractitioner !
